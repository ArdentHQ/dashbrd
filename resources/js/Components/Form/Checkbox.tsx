import cn from "classnames";

interface Properties extends React.InputHTMLAttributes<HTMLInputElement> {
    isInvalid?: boolean;
}

export const Checkbox = ({ className, isInvalid = false, ...properties }: Properties): JSX.Element => (
    <input
        data-testid="Checkbox"
        type="checkbox"
        className={cn(
            "dark:form-checkbox-dark form-checkbox box-border h-5 w-5 rounded border-2 text-theme-primary-600 transition duration-100 ease-in",
            isInvalid
                ? "border-theme-danger-400"
                : "border-theme-secondary-300 enabled:hover:border-theme-primary-400 enabled:checked:hover:border-theme-primary-400 dark:enabled:border-theme-dark-200 dark:enabled:bg-theme-dark-900 dark:enabled:checked:border-theme-primary-400 dark:enabled:checked:bg-theme-primary-400 dark:enabled:hover:border-theme-primary-600 dark:enabled:hover:bg-theme-dark-900 dark:enabled:checked:hover:border-theme-primary-600 dark:enabled:checked:hover:bg-theme-primary-600",
            "disabled:border-theme-secondary-200 disabled:bg-theme-secondary-100 disabled:checked:border-theme-secondary-100 disabled:hover:bg-theme-secondary-100 dark:disabled:border-theme-dark-400 dark:disabled:bg-theme-dark-700 dark:disabled:checked:border-theme-dark-600 dark:disabled:checked:bg-theme-dark-600",
            "focus:ring-0 focus:ring-transparent focus:ring-offset-0",
            "focus-visible:ring-4 focus-visible:ring-theme-primary-300 dark:focus-visible:ring-theme-primary-700",
            className,
        )}
        {...properties}
    />
);
