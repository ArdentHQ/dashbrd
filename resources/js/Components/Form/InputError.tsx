import cn from "classnames";

interface Properties {
    message?: string;
    className?: string;
}

export const InputError = ({ message, className }: Properties): JSX.Element => {
    if (message === undefined || message === "") {
        return <></>;
    }

    return <p className={cn("text-red-600 text-sm", className)}>{message}</p>;
};
