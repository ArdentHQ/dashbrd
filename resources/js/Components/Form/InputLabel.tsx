import cn from "classnames";

interface Properties extends React.LabelHTMLAttributes<HTMLLabelElement> {
    value?: string;
    children?: React.ReactNode;
}

export const InputLabel = ({ value, className, children, ...properties }: Properties): JSX.Element => (
    <label
        data-testid="InputLabel"
        className={cn(
            "block text-sm font-medium leading-6 text-theme-secondary-700 dark:text-theme-dark-200",
            className,
        )}
        {...properties}
    >
        {value ?? children}
    </label>
);
