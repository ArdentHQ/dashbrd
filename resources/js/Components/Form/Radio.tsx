import cn from "classnames";
import { forwardRef } from "react";

interface Properties extends React.InputHTMLAttributes<HTMLInputElement> {
    isInvalid?: boolean;
}

export const Radio = forwardRef<HTMLInputElement, Properties>(
    ({ className, isInvalid = false, ...properties }: Properties, reference): JSX.Element => (
        <input
            data-testid="Radio"
            type="radio"
            ref={reference}
            className={cn(
                "dark:form-radio-dark form-radio box-border h-5 w-5 border-2 text-theme-primary-600 transition duration-100 ease-in dark:text-theme-dark-200",
                isInvalid
                    ? "border-theme-danger-400 dark:bg-theme-dark-900"
                    : "border-theme-secondary-300 enabled:hover:border-theme-primary-400 enabled:checked:hover:border-theme-primary-400 dark:border-theme-dark-200 dark:bg-theme-dark-900 dark:enabled:checked:border-theme-primary-400 dark:enabled:checked:bg-theme-primary-400 dark:enabled:hover:border-theme-primary-600 dark:enabled:checked:hover:bg-theme-primary-400",
                "disabled:border-theme-secondary-200 disabled:bg-theme-secondary-100 disabled:checked:border-theme-secondary-100 disabled:hover:bg-theme-secondary-100 dark:disabled:border-theme-dark-400 dark:disabled:bg-theme-dark-700 dark:disabled:checked:border-theme-dark-600 dark:disabled:checked:bg-theme-dark-600",
                "focus:ring-0 focus:ring-transparent focus:ring-offset-0",
                "dark:enabled:focus-within:ring-3 focus-visible:ring-4 focus-visible:ring-theme-primary-300 dark:enabled:focus-within:ring-theme-primary-700",
                className,
            )}
            {...properties}
        />
    ),
);

Radio.displayName = "Radio";
