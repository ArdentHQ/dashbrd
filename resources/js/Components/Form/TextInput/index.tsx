export * from "./TextInput";
export * from "./TextInput.contracts";
