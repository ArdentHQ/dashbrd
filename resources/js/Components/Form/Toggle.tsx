import { Switch } from "@headlessui/react";
import cn from "classnames";

interface Properties {
    checked: boolean;
    onChange: (checked: boolean) => void;
    screenReaderLabel?: string;
    className?: string;
    name?: string;
    id?: string;
    disabled?: boolean;
}

export const Toggle = ({
    checked,
    onChange,
    screenReaderLabel,
    className,
    name,
    id,
    disabled = false,
}: Properties): JSX.Element => {
    const defaultLabel = checked ? "Disable" : "Enable";

    return (
        <Switch
            data-testid="Toggle"
            checked={checked}
            onChange={onChange}
            disabled={disabled}
            name={name}
            id={id}
            className={cn(
                className,
                "relative inline-flex h-6 w-11 items-center rounded-full transition duration-200 ease-in-out focus-visible:outline-none focus-visible:ring-4 focus-visible:ring-theme-primary-300 dark:focus-visible:ring-theme-primary-700",
                disabled
                    ? "bg-theme-secondary-200 dark:bg-theme-dark-700"
                    : [
                          checked
                              ? "bg-theme-primary-600 hover:bg-theme-primary-700 active:bg-theme-primary-800 dark:bg-theme-primary-400 dark:hover:bg-theme-primary-600 dark:focus-visible:bg-theme-primary-400 dark:active:bg-theme-primary-400"
                              : "bg-theme-secondary-200 hover:bg-theme-secondary-300 active:bg-theme-secondary-400 dark:bg-theme-dark-700 dark:focus-within:bg-theme-dark-700 dark:hover:bg-theme-dark-800 dark:active:bg-theme-dark-800",
                      ],
            )}
        >
            <span className="sr-only">{screenReaderLabel ?? defaultLabel}</span>

            <span
                className={cn(
                    "inline-block h-5 w-5 rounded-full shadow-sm transition duration-200 ease-in-out",
                    disabled ? "bg-theme-secondary-400 dark:bg-theme-dark-200" : "bg-white",
                    checked ? "translate-x-[1.375rem]" : "translate-x-0.5",
                )}
            />
        </Switch>
    );
};
