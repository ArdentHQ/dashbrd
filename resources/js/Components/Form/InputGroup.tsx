import cn from "classnames";

import { forwardRef, type InputHTMLAttributes, type ReactElement, type ReactNode } from "react";

import { InputLabel } from "@/Components/Form/InputLabel";
import { getFirstFocusableElement } from "@/Utils/get-focusable";
import { isTruthy } from "@/Utils/is-truthy";

type Hint = string | [string | ReactNode, string | ReactNode];

type HintPosition = "left" | "right";

interface Properties extends Omit<InputHTMLAttributes<HTMLInputElement>, "children"> {
    children: (({ hasError }: { hasError: boolean }) => React.ReactElement) | React.ReactElement;
    hint?: string | [string | ReactNode, string | ReactNode];
    hintPosition?: HintPosition;
    error?: string;
    label?: string | ReactElement;
    // In some cases, the feedback element needs to be positioned differently due
    // to stuff like having a group of inputs
    erroredFeedbackClass?: string;
}

const InputError = ({ message, className }: { message: string; className?: string }): JSX.Element => (
    <span
        className={cn(
            "-mt-3 block bg-theme-danger-100 px-4 pb-2 pt-5 text-xs font-medium text-theme-danger-600 dark:bg-theme-danger-400 dark:text-white",
            className,
        )}
    >
        {message}
    </span>
);

const InputHint = ({ message, position }: { message: Hint; position?: HintPosition }): JSX.Element => {
    if (Array.isArray(message)) {
        return (
            <span className="flex w-full justify-between rounded-b-xl bg-theme-secondary-100 px-4 py-2 text-xs font-medium text-theme-secondary-700 dark:bg-theme-dark-950 dark:text-theme-dark-200">
                <span>{message[0]}</span>
                <span>{message[1]}</span>
            </span>
        );
    }

    return (
        <span
            className={cn(
                "block w-full rounded-b-xl bg-theme-secondary-100 px-4 py-2 text-xs font-medium text-theme-secondary-700",
                {
                    "text-left": position === "left",
                    "text-right": position === "right",
                },
            )}
        >
            {message}
        </span>
    );
};

const inputGroupClickHandler: React.MouseEventHandler<HTMLDivElement> = (event): void => {
    // If target is a button, link or input, don't focus the input
    if (
        ["button", "a", "input", "textarea"].includes(
            (event.target as HTMLButtonElement | HTMLAnchorElement | HTMLInputElement).nodeName.toLowerCase(),
        )
    ) {
        return;
    }

    getFirstFocusableElement(event.currentTarget)?.focus();
};

export const InputGroup = forwardRef<HTMLDivElement, Properties>(
    (
        { children, hint, hintPosition, error, className, label, id, erroredFeedbackClass = "-mx-px", ...properties },
        reference,
    ): JSX.Element => {
        const hasError = isTruthy(error);

        const hasHint = isTruthy(hint);

        const hasLabel = isTruthy(label);

        const hasFeedback = hasError || hasHint;

        return (
            <>
                {hasLabel && <InputLabel htmlFor={id}>{label}</InputLabel>}

                <div
                    data-testid="InputGroup"
                    ref={reference}
                    onClick={inputGroupClickHandler}
                    className={cn("rounded-xl", className, {
                        "bg-theme-secondary-100 dark:bg-theme-dark-950": hasHint && !hasError,
                        "mt-1": hasLabel,
                    })}
                    {...properties}
                >
                    {typeof children === "function" ? children({ hasError }) : children}

                    {hasFeedback && (
                        <div
                            data-testid="InputGroup__feedback"
                            className={cn({
                                [erroredFeedbackClass]: hasError,
                            })}
                        >
                            {hasError && (
                                <InputError
                                    message={error}
                                    className={cn({
                                        "rounded-b-xl": !hasHint,
                                    })}
                                />
                            )}

                            {hasHint && (
                                <InputHint
                                    message={hint}
                                    position={hintPosition}
                                />
                            )}
                        </div>
                    )}
                </div>
            </>
        );
    },
);

InputGroup.displayName = "InputGroup";
