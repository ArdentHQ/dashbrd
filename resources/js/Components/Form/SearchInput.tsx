import classNames from "classnames";
import { type HTMLAttributes } from "react";
import { TextInput } from "@/Components/Form/TextInput";
import { Icon } from "@/Components/Icon";

interface Properties extends Omit<HTMLAttributes<HTMLDivElement>, "onChange"> {
    placeholder: string;
    onChange: (query: string) => void;
    query: string;
    disabled?: boolean;
}

export const SearchInput = ({
    query,
    onChange,
    placeholder,
    disabled = false,
    className,
    ...properties
}: Properties): JSX.Element => (
    <div
        className={classNames("relative", className)}
        {...properties}
    >
        <TextInput
            data-testid="NftCollectionSlider__search"
            value={query}
            onChange={(event: React.ChangeEvent<HTMLInputElement>) => {
                onChange(event.target.value);
            }}
            placeholder={placeholder}
            className="pr-16"
            disabled={disabled}
        />

        {query !== "" && (
            <button
                data-testid="NftCollectionSlider__clear-search"
                type="button"
                className="absolute right-0 top-0 mr-14 mt-4"
                onClick={() => {
                    onChange("");
                }}
            >
                <Icon
                    name="X"
                    size="md"
                    className="text-theme-primary-900 dark:text-theme-dark-200"
                />
            </button>
        )}

        <Icon
            name="MagnifyingGlass"
            size="md"
            className={classNames("pointer-events-none absolute right-0 top-0 mr-4 mt-4", {
                "text-theme-primary-900 dark:text-theme-dark-200": !disabled,
                "text-theme-secondary-500 dark:text-theme-dark-500": disabled,
            })}
        />
    </div>
);
