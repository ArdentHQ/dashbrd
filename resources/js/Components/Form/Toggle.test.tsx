import React from "react";
import { Toggle } from "@/Components/Form/Toggle";
import { render, screen } from "@/Tests/testing-library";

describe("Toggle", () => {
    it("should render as checked", () => {
        const { asFragment } = render(
            <Toggle
                checked
                onChange={vi.fn()}
            />,
        );

        expect(screen.getByTestId("Toggle")).toBeInTheDocument();
        expect(asFragment()).toMatchSnapshot();
    });

    it("should render as unchecked", () => {
        const { asFragment } = render(
            <Toggle
                checked={false}
                onChange={vi.fn()}
            />,
        );

        expect(screen.getByTestId("Toggle")).toBeInTheDocument();
        expect(asFragment()).toMatchSnapshot();
    });

    it("should render as disabled", () => {
        const { asFragment } = render(
            <Toggle
                disabled
                checked
                onChange={vi.fn()}
            />,
        );

        expect(screen.getByTestId("Toggle")).toBeInTheDocument();
        expect(asFragment()).toMatchSnapshot();
    });
});
