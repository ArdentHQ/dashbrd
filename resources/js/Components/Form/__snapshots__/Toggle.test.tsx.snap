// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Toggle > should render as checked 1`] = `
<DocumentFragment>
  <button
    aria-checked="true"
    class="relative inline-flex h-6 w-11 items-center rounded-full transition duration-200 ease-in-out focus-visible:outline-none focus-visible:ring-4 focus-visible:ring-theme-primary-300 bg-theme-primary-600 hover:bg-theme-primary-700 active:bg-theme-primary-800"
    data-headlessui-state="checked"
    data-testid="Toggle"
    id="headlessui-switch-:r0:"
    role="switch"
    tabindex="0"
    type="button"
  >
    <span
      class="sr-only"
    >
      Disable
    </span>
    <span
      class="inline-block h-5 w-5 rounded-full shadow-sm transition duration-200 ease-in-out bg-white translate-x-[1.375rem]"
    />
  </button>
</DocumentFragment>
`;

exports[`Toggle > should render as disabled 1`] = `
<DocumentFragment>
  <button
    aria-checked="true"
    class="relative inline-flex h-6 w-11 items-center rounded-full transition duration-200 ease-in-out focus-visible:outline-none focus-visible:ring-4 focus-visible:ring-theme-primary-300 bg-theme-secondary-200"
    data-headlessui-state="checked"
    data-testid="Toggle"
    disabled=""
    id="headlessui-switch-:r2:"
    role="switch"
    tabindex="0"
    type="button"
  >
    <span
      class="sr-only"
    >
      Disable
    </span>
    <span
      class="inline-block h-5 w-5 rounded-full shadow-sm transition duration-200 ease-in-out bg-theme-secondary-400 translate-x-[1.375rem]"
    />
  </button>
</DocumentFragment>
`;

exports[`Toggle > should render as unchecked 1`] = `
<DocumentFragment>
  <button
    aria-checked="false"
    class="relative inline-flex h-6 w-11 items-center rounded-full transition duration-200 ease-in-out focus-visible:outline-none focus-visible:ring-4 focus-visible:ring-theme-primary-300 bg-theme-secondary-200 hover:bg-theme-secondary-300 active:bg-theme-secondary-400"
    data-headlessui-state=""
    data-testid="Toggle"
    id="headlessui-switch-:r1:"
    role="switch"
    tabindex="0"
    type="button"
  >
    <span
      class="sr-only"
    >
      Enable
    </span>
    <span
      class="inline-block h-5 w-5 rounded-full shadow-sm transition duration-200 ease-in-out bg-white translate-x-0.5"
    />
  </button>
</DocumentFragment>
`;
